### ISSUE TITLE
Changes to Video's Color

### ISSUE THREAD

**Issue by Animeke** [n00mkrad/flowframes#157]
Is there any settings or way to fix this issue? Is this a bug?
![Original](https://user-images.githubusercontent.com/38185410/176110871-0bf85f34-afb2-46cf-bcf2-7385cae5504a.png)

**Comment by mirh** [comment]
https://github.com/n00mkrad/flowframes/issues/123
It's absolutely a bug. But the culprit is ffmpeg technically.

**Comment by NoticeableScout121** [comment]
> #123 It's absolutely a bug. But the culprit is ffmpeg technically.

Do you know of any fix or workaround?

**Comment by neojam** [comment]
Yep, been a bug in ffmpeg since forever.
FFmpeg did always screw up colors on frame extraction to jpg/png.
At least for any of yuv420p sources, even if those are tagged correctly, like:
pix_fmt=yuv420p
color_range=tv
color_space=bt709
color_transfer=bt709
color_primaries=bt709

**Comment by neojam** [comment]
Not from my experience.
It's not as bad as it is the case with JPEGs, but still, the exported PNGs never match the source.
What ffmpeg settings do you use to extract to PNG? 

From my experience exporting to PNG from ffmpeg darkens the image and crushes blacks.
Below is an example of SD-Video (1st frame only) followed by png/jpg of that frame that were exported from:
Premiere to PNG 
Photoshop to PNG 
ffmpeg to PNG 
ffmpeg to JPG
(I attached the video source and all exported images, if you want to do your own tests.)

Exported PNGs from Photoshop and Premiere match the source video exactly. You can test it by opening both, the source video (pause on 1st frame) and PNG in ffplay, so they're on top of each other to check for differences.
The colors of PNG that was created with ffmpeg wont match the source (darker colors + crushed blacks)

https://user-images.githubusercontent.com/8310378/176689899-d3659f91-f7ea-4495-a997-9bf694b8bae0.mp4

[compare2.zip](https://github.com/n00mkrad/flowframes/files/9020220/compare2.zip)

pix_fmt=yuv420p
level=31
color_range=tv
color_space=bt709
color_transfer=bt709
color_primaries=bt709
chroma_location=left
field_order=progressive

**Comment by mirh** [comment]
Sigh... Am I going to assume this is https://trac.ffmpeg.org/ticket/9672?
Or is the sRGB chunk/handling unrelated to ICC?

**Comment by ValZapod** [comment]
> sRGB chunk/handling unrelated to ICC?

Unrelated.

**Comment by mirh** [comment]
So.. then.. am I correct in believing there's no bug report about this issue on the tracker?

**Comment by neojam** [comment]
I honestly don't know it any different. FFmpeg was behaving for me like that for years.
On bright HD scenes the color changes are not as visible, but for dark SD scenes, especially badly compressed ones its quite obvious.

**Comment by mirh** [comment]
If your issue is a big SD/HD difference, that must necessarily be the lack of [tagging](https://trac.ffmpeg.org/ticket/9167) that I hinted from my first post. 
You may have been headbanging while trying to understand it, because:
 * a lot of professional editing tools reads png gAMA tags, which are screwed in ffmpeg (unclear where it's getting actually tracked, but hey, thankfully [reading](https://trac.ffmpeg.org/ticket/9673) them doesn't work either!)
 * flowframes tries to put up some quick smarty pants, not understanding the actual abomination beneath

TL;DR without any specific option, ffmpeg internally defaults to bt.601 without telling anybody (i.e. writing it down in metadata). 
On top and besides of that though, we have much more:
 * YUV formats (e.g. jpeg) are [fucked](https://trac.ffmpeg.org/ticket/9693) OOTB when they differ in color range
 * gamma is [happily left](https://trac.ffmpeg.org/ticket/9772#comment:5) to 2.2 despite whatever the output may be (I guess this is the minuscule tint that I still perceive once I have addressed everything else in this list.. or maybe they are the [other](https://trac.ffmpeg.org/ticket/3345) [little](https://trac.ffmpeg.org/ticket/3801) [nitpicks](https://trac.ffmpeg.org/ticket/7978))

**Comment by ValZapod** [comment]
Yes, there is none. I wanted to open but did not have time. Essentially
BT.709 is supposed to be using pure 2.4 gamma EOTF on perfect OLED screen
in the dark room. Yet they just invert OETF which is 1.9608.

чт, 30 июн. 2022 г., 23:08 mirh ***@***.***>:

**Comment by ValZapod** [comment]
>Yes, there is none. I wanted to open but did not have time. Essentially

I opened: https://trac.ffmpeg.org/ticket/10000

**Comment by mirh** [comment]
Cool. 
Now there's only half a dozen of bugs to fix to go /s

**Comment by ValZapod** [comment]
> Cool. Now there's only half a dozen of bugs to fix to go /s

Well, that BUG is technically fixed, because we added support for cICP in png, that has higher priority than gAMA chunk. So that means it now writes image with cICP OETF BT.709, that will show up correctly as 2.4 gamma in mpv, e.g., or even in ffmpeg.
